### ISSUE TITLE
Add log-driver for logstash

### ISSUE THREAD

**Issue by Alexandre-io** [docker/docker#14949]
Hi,

A plan to support logstash as a log-driver ?

**Comment by cpuguy83** [comment]
Logging drivers are provided by the community, by someone who has a need for it.
That said doesn't logstash support both journald and syslog?

**Comment by Alexandre-io** [comment]
Syslog works like a charm with logstash but the current driver have some security concerns like ssl (auth) and queuing, no ?

**Comment by cpuguy83** [comment]
I've updated the title for other people coming by who may want to implement this.

**Comment by rwincewicz** [comment]
I've been looking at a logging driver that outputs to RabbitMQ as that's how I'm handling the logs from the rest of my services and applications. I've got logstash picking up the messages at the other end and putting them into elasticsearch. At the moment I'm not using SSL but that's next on my to-do list but it does satisfy your queuing requirement.
It's early days so it's rough around the edges but feel free to take a look: https://github.com/rwincewicz/docker/tree/add_amqp_logdriver/daemon/logger/amqp

**Comment by aishraj** [comment]
#dibs

I'd to work on the above issue.

**Comment by runcom** [comment]
I fear the next one will be elastisearch itself

**Comment by mkuzmin** [comment]
Related issue: https://github.com/elastic/libbeat/issues/37

**Comment by aishraj** [comment]
Thanks for linking into the [libbeat issue](https://github.com/docker/docker/issues/14949#issuecomment-158766019).
Based on the comments<sup>1</sup> there, it looks like the Elastic community feels that with Docker moving to a non-embedded plugin approach via Rest API for plugin registration, it would not be a good timing to implement this. Please let me know if that is the case, and if so, the best way to get started with the new API based plugin approach.

[1] https://github.com/elastic/libbeat/issues/37#issuecomment-153944581

**Comment by bitsofinfo** [comment]
+1 using syslog is not a great match with multilined logs... logstash driver would be great, that writes to a logstash TCP input on the destination

**Comment by thaJeztah** [comment]
Also note this proposal; https://github.com/docker/docker/pull/18001

**Comment by superdump** [comment]
TCP+TLS and being able to either write to logstash or directly to elasticsearch would be good.

**Comment by mrkschan** [comment]
@aishraj my comment on libbeat https://github.com/elastic/libbeat/issues/37#issuecomment-153944581 does not represent the views from elastic nor docker nor dockerbeat.

I think we better ask if docker could improve the log driver to a non-embedded one. @cpuguy83 do you have any insight there?

**Comment by cpuguy83** [comment]
@bitsofinfo docker's loggers only deals with single lines at a time in any case, it doesn't matter what the driver is capable of.

@mrkschan I don't think the existing plugin API is a good one for logging.
It uses HTTP, which itself has tremendous overhead, it's not good for streaming, etc.
I think the way this will end up being pluggable (beyond support from golang for dynamically loading plugins) is #18001.

TLDR on #18001; Give docker a socket to connect to (udp, tcp, unix, doesn't matter) and docker will stream logs to it in it's native format that people can write adapters for.

**Comment by cpuguy83** [comment]
That said I'm not sure if the other maintainers want to also bring on a logstash driver or not.
My personal concern is vendoring more and more dependencies and bloating the code base.

**Comment by thaJeztah** [comment]
Yes I think we should try to work on supporting external plugins for logging if possible; agreed on the vendoring, and bloating the code base

**Comment by Alexandre-io** [comment]
Since the release 1.10.0 syslog over TCP+TLS were added. So my last concern is about the queuing

**Comment by Alexandre-io** [comment]
Since the release 1.10.0 syslog over TCP+TLS were added. So my last concern is about the queuing.

**Comment by rchicoli** [comment]
This is already supported by gelf log driver.  
You just need make sure to use in logstash the plugin gelf as input

```bash
gelf {
          port => 5002
          type => "gelf"
          tags => "docker"
}
```

**Comment by danielmotaleite** [comment]
with gelf, if i stop logstash for a few minutes, it loses the messages during that time... queuing messages only seem to work with lost packets, not lost of connection

**Comment by alkuzad** [comment]
@rchicoli UDP is not a choice for server logs, it can be acceptable for statistics logs but not critical production logs when we have to ensure every event is logged. Right now I do not see

**Comment by rchicoli** [comment]
@alkuzad I do agree. Btw I am willing to contribute to this, but the maintainers are not accepting any new log-drivers anymore (topic related https://github.com/docker/docker/issues/25694) and I don't really know how to rewrite this to an external plugin. I wish we had a couple of more examples or some documentation about writting external plugins. Otherwise I kind need some help to do so.

**Comment by cpuguy83** [comment]
Working on a plugin implementation for log drivers right now, hope to have a PoC by EOD.

**Comment by tyanko1** [comment]
@cpuguy83 Hey, what is the status on the logstash plugin? Could really use this!!! Thanks!

**Comment by cpuguy83** [comment]
@tyanko1 Implementation is here: #28403

Still a few more things to deal with, but it works, I have a test logging plugin up on hub (See integration-cli test from the mentioned PR).
It supports both writing and reading, when the plugin specifies that it supports reading.

**Comment by cpuguy83** [comment]
Log plugin support is now merged and should be part of 17.05.
I do not think we'll be accepting any new built-in log drivers.

Closing this. A logstash driver should be implemented as a plugin at this point. If anyone wants to discuss this feel free to ping me on Slack, or come see me @ DockerCon!

**Comment by rchicoli** [comment]
here is a `POC` of docker log plugin for logstash [docker-log-logstash](https://github.com/rchicoli/docker-log-logstash). Although it is not much better than gelf at the moment.

This is just the first alpha release. There are some planning for adding caching, buffering, etc. Though it will take awhile until it is stable and ready for production, because I am doing this beside my real work.

Let me know, if you are still interested on it. Feel free to contribute also, if you are feeling like.

### How to use:

Install
```
docker plugin install rchicoli/docker-log-logstash:0.5.1 --alias logstash
```
Logstash config
```
input {
    tcp {
            port => 5000
            codec => json
    }
}
output {
    stdout { codec => rubydebug }
}
```

Test

```
docker run --rm  -ti \
    --log-driver logstash \
    --log-opt logstash-url=127.0.0.1:5000 \
        alpine echo this is another logging driver
```
This is more an implementation of a `TCP client`, than `Logstash`.

Any ideas can be tracked on the project. For example, adding support to different input plugins, etc.

**Comment by rchicoli** [comment]
The current status of `docker-log-logstash:0.0.5` plugin. Even though it is in alpha, this plugin is capable of:
  * reconnecting to logstash server, in case of lost connection
  * caching messages to the filesystem, while logstash is down
  * send cached log information to logstash, when it is online
